### ISSUE TITLE
apt hangs during install of macfanctld

### ISSUE THREAD

**Issue by candlerb** [ansible/ansible#8163]
##### Issue Type:

Bug Report
##### Ansible Version:

1.5.4 and 1.6.6
##### Environment:

Ubuntu 14.04 server amd64
##### Summary:

ansible can hang if the remote command it is running forks a child which inherits stdout or stderr and keeps it open
##### Steps To Reproduce:

```
- hosts: localhost
  tasks:
  - apt: pkg=macfanctld state=absent purge=yes
  - apt: pkg=macfanctld state=present
```
##### Expected Results:

That the command would complete.
##### Actual Results:

The command hangs. At this point, opening another session and running ps shows:

```
root     11471  0.0  0.0  64956  2112 pts/1    S+   13:16   0:00 sudo ansible-playbook tmp.yml
root     11472  2.1  0.1  85664 17160 pts/1    S+   13:16   0:00 /usr/bin/python /usr/bin/ansible-playbook tmp.yml
root     11751  0.0  0.0   4444   648 pts/1    S+   13:16   0:00 /bin/sh -c /usr/bin/python /home/nsrc/.ansible/tmp/ansible-tmp-1405516587.94-157111181416778/apt; rm -rf /home/nsrc/.ansible/tmp/ansible-tmp-1405516587.94-157111181416778/ >/dev/null 2>&1
root     11752  9.7  0.4 120508 71432 pts/1    S+   13:16   0:00 /usr/bin/python /home/nsrc/.ansible/tmp/ansible-tmp-1405516587.94-157111181416778/apt
root     11756 13.2  0.0      0     0 pts/1    Z+   13:16   0:00 [apt-get] <defunct>
root     12000  0.5  0.0   4616   788 ?        Ss   13:16   0:00 /usr/sbin/macfanctld
```

Note that:
- apt-get has terminated, and is waiting to be reaped
- manually killing the `/usr/sbin/macfanctld` daemon allows ansible to terminate successfully

It would appear that [macfanctld is holding stderr open](https://bugs.launchpad.net/ubuntu/+source/macfanctld/+bug/1342740), as can be demonstrated like this at the shell:

```
# ( service macfanctld start 2>&1 ) | cat
 * Starting Apple MacBook (Pro) fan control daemon macfanctld
   ...done.
<< hangs at this point >>
```

IMO ansible should defend itself against this, e.g. by detecting that the child has terminated, waiting a second or two to check no more data is coming on the stdout/stderr pipes, and then closing them.

**Comment by candlerb** [comment]
Package bug report: https://bugs.launchpad.net/ubuntu/+source/macfanctld/+bug/1342740

**Comment by mpdehaan** [comment]
We do auto-daemonize service invocations to deal with bad init scripts (which are legion), but this one may not be avoidable, we can investigate the extra-daemonization code.

However, I noticed you reported Ansible 1.5.4, which is definitely a much older release.  Can you please confirm if you can reproduce this on 1.6.6?

**Comment by candlerb** [comment]
Yes, same with 1.6.6 from PPA.

```
root     10110  0.0  0.0  64956  2068 pts/1    S+   18:27   0:00 sudo ansible-playbook 8136.yml
root     10111  1.4  0.1  92292 17752 pts/1    S+   18:27   0:00 python /usr/bin/ansible-playbook 8136.yml
root     10488  0.0  0.0   4444   648 pts/1    S+   18:27   0:00 /bin/sh -c LC_CTYPE=en_US.UTF-8 LANG=en_US.UTF-8 /usr/bin/python /home/nsrc/.ansible/tmp/ansible-tmp-1405535237.77-37064573838305/apt; rm -rf /home/nsrc/.ansible/tmp/ansible-tmp-1405535237.77-37064573838305/ >/dev/null 2>&1
root     10489  5.8  0.4 125328 72324 pts/1    S+   18:27   0:00 /usr/bin/python /home/nsrc/.ansible/tmp/ansible-tmp-1405535237.77-37064573838305/apt
root     10493  8.0  0.0      0     0 pts/1    Z+   18:27   0:00 [apt-get] <defunct>
root     10833  0.4  0.0   4616   788 ?        Rs   18:27   0:00 /usr/sbin/macfanctld
```

**Comment by jimi-c** [comment]
This sounds like it could be related to https://github.com/ansible/ansible/issues/7452

**Comment by jimi-c** [comment]
I have merged in the above commit, which should resolve this issue. If you continue seeing problems related to this, please let us know. Thanks!

**Comment by candlerb** [comment]
Confirmed the problem is solved, tested with ansible-playbook 1.8 (devel f4d649bea0).

Thank you!
